Avoid blocking send on channels inside mutex

krisis opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
The server side implementation of ListenBucketNotification uses channels to send notification events from the origin of the event like PutObject to the ListenBucketHandler (used by mc find --watch internally). The channels are stored in a map which is protected by a mutex. ListenBucketHandler does the following when a client closes connection.

- sendBucketNotification returns immediately after a failed write (over HTTP). This means we have a channel that has a sender with no receiver
- removes reference to the channel from the map
- closes the channel.

On the other hand, a concurrent PutObjectHandler does the following,

- send notification event on channels read from the map (note: this is inside the critical section)
- release the namespace mutex lock

In this scenario, PutObjectHandler is blocked, holding the mutex protecting the map of channels and the namespace mutex since the channels used to send notification is blocked (remember, it has no receiver).

This change removes the cyclic dependency between the event channel and map of ARN to channels by using a separate done channel to signal that the client has quit.

Fixes #5023 
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
https://github.com/minio/minio/issues/5023
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Not yet.
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

harshavardhana commented on cmd/event-notifier.go:
```diff
@@ -230,7 +230,14 @@ func (en *eventNotifier) SendListenerEvent(arn string, event []NotificationEvent
+		//the actual http.ResponseWriter.
+		select {
+		case ch <- event:
+		default:
```
Can we log an error here?

krisis replied:
@harshavardhana What would the error message be? What do we intend to inform our users here?

harshavardhana replied:
```
errorIf(errors.New("blocked channel"), "Unable to send the event to %s, possibly no receiver", arn) 
```
This log will be removed by moving to http.ResponseWriter.

harshavardhana commented on cmd/event-notifier.go:
```diff
@@ -230,7 +230,14 @@ func (en *eventNotifier) SendListenerEvent(arn string, event []NotificationEvent
-		ch <- event
+		// To avoid blocking of send event over the channel
+		// while inside critical section.
+		//FIXME: The intermediary channel can be replaced by
```
Also comment can have a space after `//`

donatello commented on cmd/event-notifier.go:
```diff
@@ -230,7 +230,19 @@ func (en *eventNotifier) SendListenerEvent(arn string, event []NotificationEvent
+		case ch <- event:
+		default:
+			// When there is no receiver at the other end
+			// of the event channel, the channel blocks
```
I don't think fix is correct - it would introduce a new problem. The ListenBucketNotificationHandler is the only go routine that listens on the channel and writes any received events to the http response writer. This means that between iterations of this loop, sending on the channel is blocked (because there is no listener). This in turn means that this select block will drop events, thus losing them.

See https://play.golang.org/p/oPmqFoaNA7 for a demonstration.

harshavardhana replied:
Loosing is acceptable in this scenario @donatello , we are aware that it will happen. So there is a comment on why it is okay ignore. The reasoning is simple this only happens on a unbuffered channel when indeed there is no receiver. This part of the code is also scheduled to be quickly fixed by moving to response writers instead of channel.

harshavardhana replied:
Also this situation only happens when there are other active listeners.. which are non-blocking so it is highly unlikely that all of them are blocking and not receiving. Single listener `default:` should never get executed.

donatello reviewed (changes requested):
This bug needs some rethinking - possibly directly writing to the response-writer will work, but it needs to be synchronized with other writers.

donatello commented on cmd/bucket-notification-handlers.go:
```diff
@@ -226,23 +226,14 @@ func writeNotification(w http.ResponseWriter, notification map[string][]Notifica
-		case <-time.After(globalSNSConnAlive): // Wait for global conn active seconds.
-			if err := writeNotification(w, dummyEvents); err != nil {
+		case <-time.After(globalSNSConnAlive):
+			if err := writeNotification(w, emptyEvent); err != nil {
```
It is not clear to me if `ResponseWriter` is concurrent-safe. We are writing in one thread here, and in another when events are sent.

krisis committed:
Fix ListenBucketNotification deadlock

Previously ListenBucketNotificationHandler could deadlock with
PutObjectHandler's eventNotify call when a client closes its
connection. This change removes the cyclic dependency between the
channel and map of ARN to channels by using a separate done channel to
signal that the client has quit.

donatello commented on cmd/bucket-notification-handlers.go:
```diff
@@ -226,25 +218,51 @@ func writeNotification(w http.ResponseWriter, notification map[string][]Notifica
+}
+
+// sendNotificationEvent sends notification events on the data channel
+// until doneCh is not closed
```
s/until doneCh is not closed/unless doneCh is closed/

krisis replied:
done

donatello commented on cmd/bucket-notification-handlers.go:
```diff
@@ -226,25 +218,51 @@ func writeNotification(w http.ResponseWriter, notification map[string][]Notifica
+// waitForListener writes event notification OR whitespaces on
+// ResponseWriter until client closes connection
+func (l *listenChan) waitForListener(w http.ResponseWriter) {
+	emptyEvent := map[string][]NotificationEvent{"Records": nil}
```
Can be simplified to just: 
`emptyEvent := map[string][]NotificationEvent{}`
since nil is a zero value anyway.

harshavardhana replied:
You need "Records" otherwise the client unmarshal will fail. @donatello

donatello commented on cmd/bucket-notification-handlers.go:
```diff
@@ -226,25 +218,51 @@ func writeNotification(w http.ResponseWriter, notification map[string][]Notifica
+		case <-time.After(globalSNSConnAlive):
+			if err := writeNotification(w, emptyEvent); err != nil {
+				errorIf(err, "Unable to write empty notification")
+				close(l.doneCh)
```
The close statement can be factored out from both cases and made into a defer statement at the beginning of the function.

krisis replied:
done

donatello reviewed (dismissed):
LGTM - suggesting minor changes.

harshavardhana commented on cmd/bucket-notification-handlers.go:
```diff
@@ -226,25 +218,51 @@ func writeNotification(w http.ResponseWriter, notification map[string][]Notifica
+func (l *listenChan) sendNotificationEvent(events []NotificationEvent) {
+	select {
+	// Returns immediately if receiver has quit.
+	case <-l.doneCh:
```
We can add a test case for `doneCh` here.

krisis replied:
I have added a test case.

harshavardhana commented on cmd/bucket-notification-handlers.go:
```diff
@@ -226,25 +218,51 @@ func writeNotification(w http.ResponseWriter, notification map[string][]Notifica
-				errorIf(err, "Unable to write notification to client.")
+		case <-time.After(globalSNSConnAlive):
+			if err := writeNotification(w, emptyEvent); err != nil {
+				errorIf(err, "Unable to write empty notification")
```
Perhaps this log is not needed, most probably client actually died/disconnected here.. Its a good case of us.

krisis replied:
This log captures failure of server to write to the client when it was writing empty notification events as opposed actual events like in the previous case statement. Both these error message indicate that the client has quit, but the different error messages allow us to uniquely identify the exact case.

harshavardhana replied:
Correct but client dead is a good case so we don't need to log that message, any other we can log.

krisis replied:
Both the error logs assume that any error during Write on http.ResponseWriter is due to client quitting ListenBucketNotification. Are you suggesting we should only log error values other than the `ECONNRESET` equivalent `net.OpError`?

harshavardhana replied:
Yes..any broken pipe. We can ignore where the connection is dead sort of.

harshavardhana reviewed (approved):
LGTM tested and working.

harshavardhana commented on cmd/server_test.go:
```diff
@@ -522,32 +522,6 @@ func (s *TestSuiteCommon) TestListenBucketNotificationHandler(c *check) {
-
-	// Change global value from 5 second to 100millisecond.
-	globalSNSConnAlive = 100 * time.Millisecond
-	req, err = newTestSignedRequest("GET",
```
Why not leave this here for now? @krisis - this can be fixed when the Bala's refactor change is merged.

krisis replied:
It serves no purpose. It sends a http request to the listenBucketNotication handler and nothing else. So, there is never a bad time to remove it.
